mysys/my_uuid_v7.c

```diff
@@ -0,0 +1,162 @@
+#include <my_rnd.h>
+#include <m_string.h>
+#include <myisampack.h> /* mi_int2store, mi_int4store */
+
+static my_bool my_uuid_v7_inited= 0;
+static uint64 borrowed_microseconds;
+static uint64 uuid_time= 0;
+
+static mysql_mutex_t LOCK_uuid_v7_generator;
+
+#define UUID_VERSION                         0x7000
+#define UUID_VERSION_MASK                    0x0FFF
+#define UUID_VARIANT                         0x8000000000000000
+#define UUID_VARIANT_MASK                    0x3FFFFFFFFFFFFFFF
+#define MICROSECONDS_TO_12BIT_MAPPING_FACTOR 4.096
+#define RATE_LIMIT                           500000
+#define SLEEP_MICROSECONDS                   250000
+#define SLEEP_MILLISECONDS                   250
+
+/**
+  Init structures needed for my_uuid_v7
+*/
+void my_uuid_v7_init()
+{
+  if (my_uuid_v7_inited)
+    return;
+  my_uuid_v7_inited= 1;
+  borrowed_microseconds= 0;
+
+  mysql_mutex_init(key_LOCK_uuid_v7_generator, &LOCK_uuid_v7_generator, MY_MUTEX_INIT_FAST);
+}
+
+/**
+   Create a global unique identifier version 7(uuidv7)
+
+   @func  my_uuid_v7()
+   @param to   Store uuidv7 here. Must be of size MY_UUID_SIZE (16)
+*/
+void my_uuid_v7(uchar *to)
+{
+  uint64 tv, unix_ts_ms;
+  uint16 sub_ms_precision, version_and_sub_ms_precision;
+  uint64 rand, variant_and_rand;
+
+  DBUG_ASSERT(my_uuid_v7_inited);
+
+  tv= my_hrtime().val;
+
+  // Regulate the access to uuid_time and microseconds
+  mysql_mutex_lock(&LOCK_uuid_v7_generator);
+
+  if (likely(tv > uuid_time))
+  {
+    /*
+      Current time is ahead of last timestamp, as it should be.
+      If we "borrowed time", give it back, just as long as we
+      stay ahead of the previous timestamp.
+    */
+    if (borrowed_microseconds)
+    {
+      uint64 delta;
+      /*
+        -1 so we won't make tv= uuid_time for microseconds >= (tv - uuid_time)
+      */
+      delta= MY_MIN(borrowed_microseconds, (uint16)(tv - uuid_time -1));
+      tv-= delta;
+      borrowed_microseconds-= delta;
+    }
+  }
+  else if (unlikely(tv <= uuid_time))
+  {
+    /*
+      If several requests for UUIDs end up on the same tick
+      or if the system clock is turned *back*.
+      We add milli-seconds to make them different.
+      ( current_timestamp + microseconds * calls_in_this_period )
+      may end up > next_timestamp; this is OK. Nonetheless, we'll
+      try to unwind microseconds when we get a chance to.
+    */
+    borrowed_microseconds += uuid_time - tv + 1;
+    tv = uuid_time + 1;
+
+    if (borrowed_microseconds > RATE_LIMIT)
+    {
+      /* We are building up too much borrowed time, >500 milliseconds.
+         The output could become non-time-sortable if the server
+         process restarts, causing `borrowed_microseconds` to reset to 0.
+
+         Sleep until we have HALF of that maximum debt,
+         in order to avoid repeatedly blocking and sleeping
+         on successive calls.
+      */
+      my_sleep(SLEEP_MILLISECONDS);
+      borrowed_microseconds -= SLEEP_MICROSECONDS;
```

Comment:
Naming nitpick: I would rather see `RATE_LIMIT` named something like `MAX_BORROWED_MICROSECONDS`, and just use `MAX_BORROWED_MICROSECONDS / 2` instead of having a second `#define` which could get out of sync out of time.